planner/core/rule_groupby_str2int.go

```diff
@@ -0,0 +1,388 @@
+	newAgg := LogicalAggregation{
+		AggFuncs:     make([]*aggregation.AggFuncDesc, 0, len(agg.AggFuncs)),
+		GroupByItems: make([]expression.Expression, 0, len(agg.GroupByItems)),
+		aggHints:     agg.aggHints,
+	}.Init(agg.ctx, agg.blockOffset)
+
+	for _, aggfunc := range agg.AggFuncs {
+		newAf := aggfunc.Clone()
+		for i, arg := range newAf.Args {
+			if arg.Equal(agg.ctx, groupKey) {
+				if i != 0 || newAf.Name != "firstrow" {
+					return nil
+				}
+				newAf.Args[i] = newCol
+				newAf.RetTp = types.NewFieldType(mysql.TypeLong)
+			}
+		}
+		newAgg.AggFuncs = append(newAgg.AggFuncs, newAf)
+	}
+
+	for _, item := range agg.GroupByItems {
+		if item.Equal(agg.ctx, groupKey) {
+			newAgg.GroupByItems = append(newAgg.GroupByItems, newCol)
+		} else {
+			newAgg.GroupByItems = append(newAgg.GroupByItems, item)
+		}
+	}
+
+	aggSchema := make([]*expression.Column, 0, agg.Schema().Len())
+	for _, column := range agg.Schema().Columns {
+		if groupKey.Equal(agg.ctx, column) {
+			aggSchema = append(aggSchema, newCol)
+		} else {
+			aggSchema = append(aggSchema, column)
+		}
+	}
+	newAgg.SetSchema(expression.NewSchema(aggSchema...))
+	newAgg.SetChildren(proj)
+	return newAgg
+}
+
+// createProjectOnAgg creates a Project on top of Aggregate.
+// That Project will translate int back to string.
+func createProjectOnAgg(newAgg *LogicalAggregation,
+	groupKey expression.Expression, caseExpr expression.Expression) (*LogicalProjection, *expression.Column) {
+	found := false
+	proj := LogicalProjection{
+		Exprs: make([]expression.Expression, 0, newAgg.Schema().Len()),
+	}.Init(newAgg.ctx, newAgg.blockOffset)
+
+	newCol := &expression.Column{
+		UniqueID: newAgg.SCtx().GetSessionVars().AllocPlanColumnID(),
+		RetType:  caseExpr.GetType(),
+	}
+
+	projSchema := make([]*expression.Column, 0, newAgg.Schema().Len())
+
+	for _, column := range newAgg.Schema().Columns {
+		// if the agg key is a column
+		if column.Equal(newAgg.ctx, groupKey) {
+			proj.Exprs = append(proj.Exprs, caseExpr)
+			projSchema = append(projSchema, newCol)
+			found = true
+		} else {
+			proj.Exprs = append(proj.Exprs, column)
+			projSchema = append(projSchema, column)
+		}
+	}
+
+	if !found {
+		return nil, nil
+	}
+
+	proj.SetSchema(expression.NewSchema(projSchema...))
+	proj.SetChildren(newAgg)
+	return proj, newCol
+}
+
+func (a *groupbyStr2IntConverter) recursiveOptimize(ctx context.Context, p LogicalPlan,
+	mapping map[string]*expression.Column, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	newChildren := make([]LogicalPlan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		newChild, err := a.recursiveOptimize(ctx, child, mapping, opt)
+		if err != nil {
+			return nil, err
+		}
+		newChildren = append(newChildren, newChild)
+	}
+	p.SetChildren(newChildren...)
+
+	// we have to update the parent's reference to the child columns
+	p.replaceExprColumns(mapping)
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, mapping)
+	}
```

Comment:
I'm not sure, but it seems like we can just invoke `p.replaceExprColumns(mapping)` once in `groupbyStr2IntConverter.optimize:L383` after finishing this optimization.

Reply:
The method only changes itself, not child operators. And it doesn't change schema, which we have to do separately.

Reply:
Oh you are right, `plan.replaceExprColumns` doesn't involve its children recursively.